server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java

```diff
@@ -140,13 +142,27 @@ public void execute(SearchContext context) {
         }
 
         try {
+            int[] sortedDocIds = Arrays.copyOfRange(context.docIdsToLoad(), context.docIdsToLoadFrom(), context.docIdsToLoadSize());
+            Arrays.sort(sortedDocIds);
+
+            // preserve the original order of hits in inverted index
+            Map<Integer, ArrayList<Integer>> docIdToIndex = new HashMap<>();
```

Comment:
I wonder if it'd be better to use a static inner class and a custom comparator ?

Reply:
It seems better to preserve the original order in a custom comparator than using a hash map. I am going to update it.